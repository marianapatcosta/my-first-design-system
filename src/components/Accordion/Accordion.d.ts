import React, { CSSProperties, ReactNode } from 'react';
export interface AccordionProps {
    /**
     * element id
     */
    id?: string;
    /**
     * additional css class
     */
    className?: string;
    /**
     * additional style
     */
    style?: CSSProperties;
    /**
     * Accordion's title
     */
    title: string | ReactNode;
    /**
     * Accordion's content; is only visible when the component is expanded
     */
    content?: string | ReactNode;
    /**
     * Is Accordion disabled?
     */
    disabled?: boolean;
}
export declare const Accordion: React.FC<AccordionProps>;
