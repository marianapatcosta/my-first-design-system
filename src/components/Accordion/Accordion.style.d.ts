export declare const StyledAccordion: import("styled-components").StyledComponent<"div", any, {}, never>;
export declare const StyledAccordionTitle: import("styled-components").StyledComponent<"div", any, {
    isExpanded?: boolean | undefined;
    disabled?: boolean | undefined;
}, never>;
export declare const StyledAccordionArrow: import("styled-components").StyledComponent<"img", any, {
    isExpanded?: boolean | undefined;
}, never>;
export declare const StyledAccordionContent: import("styled-components").StyledComponent<"div", any, {
    isExpanded?: boolean | undefined;
}, never>;
