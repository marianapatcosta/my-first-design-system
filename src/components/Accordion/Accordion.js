var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { useEffect, useRef, useState, } from 'react';
import { Chevron } from '../../assets/icons';
import { isEventValid } from '../../utils/utils';
import { StyledAccordion, StyledAccordionTitle, StyledAccordionArrow, StyledAccordionContent, } from './Accordion.style';
export var Accordion = function (_a) {
    var title = _a.title, content = _a.content, disabled = _a.disabled, props = __rest(_a, ["title", "content", "disabled"]);
    var _b = useState(false), isExpanded = _b[0], setIsExpanded = _b[1];
    var accordionRef = useRef(null);
    useEffect(function () {
        var handleClickOutside = function (event) {
            var element = event.target;
            if (accordionRef.current && !(accordionRef === null || accordionRef === void 0 ? void 0 : accordionRef.current.contains(element))) {
                event.preventDefault();
                setIsExpanded(false);
            }
        };
        document.addEventListener('mouseup', handleClickOutside);
        return function () { return document.removeEventListener('mouseup', handleClickOutside); };
    }, [accordionRef]);
    var toggleIsExpanded = function (event) {
        return isEventValid(event) &&
            !disabled &&
            setIsExpanded(function (prevIsExpanded) { return !prevIsExpanded; });
    };
    return (_jsxs(StyledAccordion, __assign({}, props, { ref: accordionRef }, { children: [_jsxs(StyledAccordionTitle, __assign({ role: 'button', tabIndex: disabled ? -1 : 0, "aria-disabled": disabled, "aria-expanded": isExpanded, onClick: toggleIsExpanded, onKeyDown: toggleIsExpanded, disabled: disabled, isExpanded: isExpanded }, { children: [title, _jsx(StyledAccordionArrow, { src: Chevron, alt: 'chevron-icon', isExpanded: isExpanded }, void 0)] }), void 0),
            _jsx(StyledAccordionContent, __assign({ isExpanded: isExpanded }, { children: content }), void 0)] }), void 0));
};
Accordion.defaultProps = {
    id: '',
    className: '',
    style: {},
    disabled: false,
    title: '',
    content: '',
};
