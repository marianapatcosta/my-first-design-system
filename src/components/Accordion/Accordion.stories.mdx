import { Meta } from '@storybook/addon-docs/blocks';
import { Accordion } from './Accordion';
import Test from '../../assets/images/test.jpg';

<Meta title='Design System|Accordion' component={Accordion} />

# Accordion

Accordion component

## Usage

Use the Accordion when you need to show a title and the content should only be visible on demand,
by clicking on the component. Both title and content can be string or another components/ HTML elements

- Default Accordion:
  <Story name='Default'>
    <Accordion
      title='Accordion Title'
      content={
        <div>
          Accordion content <img style={{ width: '200px' }} src={Test} />
        </div>
      }
    />
  </Story>

## Other important notes

not applicable.
