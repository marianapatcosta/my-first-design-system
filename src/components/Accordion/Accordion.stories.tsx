import React from 'react';
import { Story, Meta } from '@storybook/react/types-6-0';
import { Accordion, AccordionProps } from './Accordion';
import Test from '../../assets/images/test.jpg';

export default {
  title: 'Accordion',
  component: Accordion,
  args: {
    title: 'Accordion title',
    content: (
      <div>
        Accordion content <img style={{ width: '200px' }} src={Test} alt='test' />
      </div>
    ),
    style: { width: '300px' },
  },
} as Meta;

const Template: Story<AccordionProps> = (args) => <Accordion {...args} />;

export const Default = Template.bind({});
Default.args = {};
