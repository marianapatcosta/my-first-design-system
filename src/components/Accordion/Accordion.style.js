var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
import styled from 'styled-components';
import { baseStyles, disabledStyles, colors } from '../../styles';
export var StyledAccordion = styled.div(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n  ", ";\n  display: flex;\n  flex-direction: column;\n  overflow: hidden;\n  max-height: 15rem;\n  max-width: 15rem;\n"], ["\n  ", ";\n  display: flex;\n  flex-direction: column;\n  overflow: hidden;\n  max-height: 15rem;\n  max-width: 15rem;\n"])), baseStyles);
export var StyledAccordionTitle = styled.div(templateObject_2 || (templateObject_2 = __makeTemplateObject(["\n  display: flex;\n  align-items: center;\n  justify-content: space-between;\n  background-color: ", ";\n  padding: 0.4rem 1rem;\n  border-radius: 0.2rem;\n  box-shadow: 0 0.0625rem 0.3rem ", ";\n  cursor: pointer;\n  z-index: 1;\n  margin: 0.125rem;\n\n  ", "\n\n  ", "\n"], ["\n  display: flex;\n  align-items: center;\n  justify-content: space-between;\n  background-color: ", ";\n  padding: 0.4rem 1rem;\n  border-radius: 0.2rem;\n  box-shadow: 0 0.0625rem 0.3rem ", ";\n  cursor: pointer;\n  z-index: 1;\n  margin: 0.125rem;\n\n  ", "\n\n  ",
    "\n"])), colors.primary, colors.shadow, function (_a) {
    var disabled = _a.disabled;
    return disabled && " " + disabledStyles;
}, function (_a) {
    var isExpanded = _a.isExpanded;
    return isExpanded &&
        "border-bottom-left-radius: 0;\n      border-bottom-right-radius: 0;";
});
export var StyledAccordionArrow = styled.img(templateObject_3 || (templateObject_3 = __makeTemplateObject(["\n  width: 0.8rem;\n  margin-left: 0.5rem;\n  ", ";\n  transition: transform 0.5s ease-in-out;\n\n  ", "\n"], ["\n  width: 0.8rem;\n  margin-left: 0.5rem;\n  ", ";\n  transition: transform 0.5s ease-in-out;\n\n  ",
    "\n"])), colors.icon, function (_a) {
    var isExpanded = _a.isExpanded;
    return isExpanded &&
        "border-bottom-left-radius: 0;\n    transform: rotate(180deg);";
});
export var StyledAccordionContent = styled.div(templateObject_4 || (templateObject_4 = __makeTemplateObject(["\n  height: 0;\n  max-height: 0;\n  border: 0.125rem solid transparent;\n  overflow: auto;\n  transition: all 0.5s ease-in-out;\n  visibility: hidden;\n  padding: 0 1rem;\n\n  -ms-overflow-style: none;\n  scrollbar-width: none;\n\n  &::-webkit-scrollbar {\n    display: none;\n  }\n  ", "\n"], ["\n  height: 0;\n  max-height: 0;\n  border: 0.125rem solid transparent;\n  overflow: auto;\n  transition: all 0.5s ease-in-out;\n  visibility: hidden;\n  padding: 0 1rem;\n\n  -ms-overflow-style: none;\n  scrollbar-width: none;\n\n  &::-webkit-scrollbar {\n    display: none;\n  }\n  ",
    "\n"])), function (_a) {
    var isExpanded = _a.isExpanded;
    return isExpanded &&
        "border-color: " + colors.primary + ";\n      border-bottom-left-radius: 0.2rem;\n      border-bottom-right-radius: 0.2rem;\n      padding: 0.4rem 1rem;\n      height: 100%;  \n      max-height: 13rem;\n      visibility: visible;\n      ";
});
var templateObject_1, templateObject_2, templateObject_3, templateObject_4;
