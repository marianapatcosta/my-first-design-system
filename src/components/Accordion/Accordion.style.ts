import styled from 'styled-components';
import { baseStyles, disabledStyles, colors } from '../../styles';

export const StyledAccordion = styled.div`
  ${baseStyles};
  display: flex;
  flex-direction: column;
  overflow: hidden;
  max-height: 15rem;
`;

export const StyledAccordionTitle = styled.div<{
  isExpanded?: boolean;
  disabled?: boolean;
}>`
  display: flex;
  align-items: center;
  justify-content: space-between;
  background-color: ${colors.primary};
  padding: 0.4rem 1rem;
  border-radius: 0.2rem;
  box-shadow: 0 0.0625rem 0.3rem ${colors.shadow};
  color: ${colors.white};
  cursor: pointer;
  z-index: 1;
  margin: 0.125rem;

  ${({ disabled }) => disabled && ` ${disabledStyles}`}

  ${({ isExpanded }) =>
    isExpanded &&
    `border-bottom-left-radius: 0;
      border-bottom-right-radius: 0;`}
`;

export const StyledAccordionArrow = styled.img<{ isExpanded?: boolean }>`
  width: 0.5rem;
  margin-left: 0.5rem;
  transform: rotate(90deg);
  filter: invert(98%) sepia(26%) saturate(2%) hue-rotate(98deg) brightness(116%)
    contrast(100%);
  transition: transform 0.5s ease-in-out;

  ${({ isExpanded }) =>
    isExpanded &&
    `border-bottom-left-radius: 0;
       transform: rotate(270deg);`}
`;

export const StyledAccordionContent = styled.div<{ isExpanded?: boolean }>`
  height: 0;
  max-height: 0;
  border: 0.125rem solid transparent;
  overflow: auto;
  transition: all 0.5s ease-in-out;
  visibility: hidden;
  padding: 0 1rem;
  
  -ms-overflow-style: none;
  scrollbar-width: none;

  &::-webkit-scrollbar {
    display: none;
  }
  ${({ isExpanded }) =>
    isExpanded &&
    `border-color: ${colors.primary};
      border-bottom-left-radius: 0.2rem;
      border-bottom-right-radius: 0.2rem;
      padding: 0.4rem 1rem;
      height: 100%;  
      max-height: 13rem;
      visibility: visible;
      box-shadow: 0 0.0625rem 0.3rem ${colors.shadow};
      `}
`;
