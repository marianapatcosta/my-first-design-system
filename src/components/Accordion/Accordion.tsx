import React, {
  CSSProperties,
  useEffect,
  useRef,
  useState,
  ReactNode,
} from 'react';
import { Chevron } from '../../assets/icons';
import { isEventValid } from '../../utils/utils';
import {
  StyledAccordion,
  StyledAccordionTitle,
  StyledAccordionArrow,
  StyledAccordionContent,
} from './Accordion.style';

export interface AccordionProps {
  /**
   * element id
   */
  id?: string;
  /**
   * additional css class
   */
  className?: string;
  /**
   * additional style
   */
  style?: CSSProperties;
  /**
   * Accordion's title
   */
  title: string | ReactNode;
  /**
   * Accordion's content; is only visible when the component is expanded
   */
  content?: string | ReactNode;
  /**
   * Is Accordion disabled?
   */
  disabled?: boolean;
}

export const Accordion: React.FC<AccordionProps> = ({
  title,
  content,
  disabled,
  ...props
}) => {
  const [isExpanded, setIsExpanded] = useState<boolean>(false);
  const accordionRef = useRef<HTMLDivElement>(null);

  useEffect(() => {
    const handleClickOutside = (event: any) => {
      const element = event.target as Node | null;

      if (accordionRef.current && !accordionRef?.current.contains(element)) {
        event.preventDefault();

        setIsExpanded(false);
      }
    };
    document.addEventListener('mouseup', handleClickOutside);

    return () => document.removeEventListener('mouseup', handleClickOutside);
  }, [accordionRef]);

  const toggleIsExpanded = (event: any) =>
    isEventValid(event) &&
    !disabled &&
    setIsExpanded((prevIsExpanded) => !prevIsExpanded);

  return (
    <StyledAccordion {...props} ref={accordionRef}>
      <StyledAccordionTitle
        role='button'
        tabIndex={disabled ? -1 : 0}
        aria-disabled={disabled}
        aria-expanded={isExpanded}
        onClick={toggleIsExpanded}
        onKeyDown={toggleIsExpanded}
        disabled={disabled}
        isExpanded={isExpanded}
      >
        {title}
        <StyledAccordionArrow
          src={Chevron}
          alt='chevron-icon'
          isExpanded={isExpanded}
        />
      </StyledAccordionTitle>

      <StyledAccordionContent isExpanded={isExpanded}>
        {content}
      </StyledAccordionContent>
    </StyledAccordion>
  );
};

Accordion.defaultProps = {
  id: '',
  className: '',
  style: {},
  disabled: false,
  title: '',
  content: '',
};
