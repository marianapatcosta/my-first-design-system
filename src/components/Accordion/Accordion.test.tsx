import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { Accordion } from './Accordion';

describe('Accordion', () => {
  const defaultProps = {
    title: 'Accordion title',
    content: 'Accordion content',
  };

  test('renders without errors and matches snapshot', () => {
    const {
      container: { firstChild },
    } = render(<Accordion {...defaultProps} />);

    expect(firstChild).toMatchSnapshot();
  });

  test('should display title', () => {
    const { getByText } = render(<Accordion {...defaultProps} />);
    getByText(/title/i);
  });

  test('should show content on click', () => {
    const { getByText } = render(<Accordion {...defaultProps} />);
    const accordionTitle = getByText(/title/i);
    const accordionContent = getByText(/content/i);
    fireEvent.click(accordionTitle);

    expect(accordionContent).toBeVisible();
  });

  test('should set isExpanded to false if click outside of accordion', () => {
    const { getByText, getByAltText } = render(<Accordion {...defaultProps} />);
    const accordion = getByText(/title/i) as HTMLDivElement;
    fireEvent.click(accordion);
    const chevron = getByAltText('chevron-icon') as HTMLImageElement;
    fireEvent.mouseUp(document);
    const style = window.getComputedStyle(chevron);
    expect(style.transform).toBe('');
  });

  describe('Disabled Accordion', () => {
    test('should display title', () => {
      const { getByText } = render(
        <Accordion {...defaultProps} disabled={true} />
      );
      getByText(/title/i);
    });

    test('should be disabled', () => {
      const { getByText } = render(<Accordion {...defaultProps} disabled />);
      const accordionTitle = getByText(/title/i);
      expect(accordionTitle).toHaveAttribute('disabled');
    });

    test('should not handle click events', () => {
      const { getByText } = render(
        <Accordion {...defaultProps} disabled={true} />
      );

      const accordionTitle = getByText(/title/i);
      const accordionContent = getByText(/content/i);
      fireEvent.click(accordionTitle);

      const titleStyle = window.getComputedStyle(accordionTitle);
      const contentStyle = window.getComputedStyle(accordionContent);
      expect(titleStyle.pointerEvents).toBe('none');
      expect(contentStyle.visibility).toBe('hidden');
    });
  });
});
