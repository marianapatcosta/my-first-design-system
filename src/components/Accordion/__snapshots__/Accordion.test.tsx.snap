// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion renders without errors and matches snapshot 1`] = `
<div
  class="sc-bdfBwQ jSFOHL"
  id=""
>
  <div
    aria-disabled="false"
    aria-expanded="false"
    class="sc-gsTCUz hkoBPB"
    role="button"
    tabindex="0"
  >
    Accordion title
    <img
      alt="chevron-icon"
      class="sc-dlfnbm blaizu"
      src="chevron.svg"
    />
  </div>
  <div
    class="sc-hKgILt dRDRsX"
  >
    Accordion content
  </div>
</div>
`;
